internal/util/logger/logger.go

```diff
@@ -0,0 +1,291 @@
+package logger
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/rs/zerolog"
+)
+
+const (
+	loggerFieldName      = "logger"
+	loggerFieldSeparator = "."
+	loggerDefaultName    = "numaplane"
+	messageFieldName     = "msg"
+	timestampFieldName   = "ts"
+)
+
+/*
+Log Level Mapping:
+| NumaLogger semantic | logr verbosity  | zerolog        |
+| ------------------- | --------------- | -------------- |
+|    fatal            |    0            |    4           |
+|    error            |    error (NA)   |    3           |
+|    warn             |    1            |    2           |
+|    info             |    2            |    1           |
+|    debug            |    3            |    0           |
+|    verbose          |    4            |    -2 (custom) |
+*/
+
+const (
+	fatalLevel   = 0
+	warnLevel    = 1
+	infoLevel    = 2
+	debugLevel   = 3
+	verboseLevel = 4
+)
+
+var logrVerbosityToZerologLevelMap = map[int]zerolog.Level{
+	fatalLevel:   4,
+	warnLevel:    2,
+	infoLevel:    1,
+	debugLevel:   0,
+	verboseLevel: -2,
+}
+
+const defaultLevel = infoLevel
+
+type loggerKey struct{}
+
+// NumaLogger is the struct containing a pointer to a logr.Logger instance.
+type NumaLogger struct {
+	LogrLogger *logr.Logger
+}
+
+// LogSink implements logr.LogSink using zerolog as base logger.
+type LogSink struct {
+	l     *zerolog.Logger
+	name  string
+	depth int
+}
+
+// New returns a new NumaLogger with a logr.Logger instance with a default setup for zerolog.
+// The writer argument sets the output the logs will be written to. If it is nil, os.Stdout will be used.
+// The level argument sets the log level value for this logger instance.
+func New(writer *io.Writer, level *int) NumaLogger {
+	// Set some zerolog customization
+	zerolog.MessageFieldName = messageFieldName
+	zerolog.TimestampFieldName = timestampFieldName
+	zerolog.TimeFieldFormat = time.RFC3339Nano
+	zerolog.LevelFieldMarshalFunc = zerologLevelFieldMarshalFunc
+	zerolog.CallerMarshalFunc = zerologCallerMarshalFunc
+
+	// Set zerolog global level to the most verbose level
+	zerolog.SetGlobalLevel(logrVerbosityToZerologLevelMap[verboseLevel])
+
+	w := io.Writer(os.Stdout)
+	if writer != nil {
+		w = *writer
+	}
+
+	lvl := int(defaultLevel)
+	if level != nil {
+		lvl = *level
```

Comment:
I don't follow how this `level` is used? and it is default to debug level if not provided?

Reply:
`level` is the argument to the function. If it is omitted, the function will use a default level (I decided to use Info level for default).